### PR TITLE
Move subnets creation and route table association to for_each

### PR DESCRIPTION
**Pull request by thomast1906**
### Change description ###
Move subnets creation and route table association to for_each
- .tfvars example, https://github.com/hmcts/aks-cft-deploy/blob/update-network-subnets/environments/network/sbox.tfvars#L9-L33

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[X] No
```

### CODE REVIEW

File: 00-init.tf [hunk 1]
```diff
@@ -1,3 +1,4 @@
 terraform {
   required_version = ">= 0.12.0"
+  experiments      = [module_variable_optional_attrs]
```
Comment (timja) [inline 00-init.tf:3]: I don't think we should use this.

https://www.terraform.io/docs/language/expressions/type-constraints.html#experimental-optional-object-type-attributes

> Until the experiment is concluded, the behavior of this feature may see breaking changes even in minor releases. We recommend using this feature only in prerelease versions of modules as long as it remains experimental.

Just don't define the type?

File: 02-inputs-optional.tf [hunk 2]
```diff
@@ -13,18 +13,6 @@ variable "route_next_hop_type" {
...
-    name                   = string
-    address_prefix         = string
-    next_hop_type          = string
-    next_hop_in_ip_address = string
-  }))
-
-  default = []
-}
-
 variable "additional_subnets" {
```
Comment (timja) [inline 02-inputs-optional.tf:16]: this should go shouldn't it?

File: 12-connectivity.tf [hunk 3]
```diff
@@ -103,12 +47,12 @@ resource "azurerm_route" "additional_route" {
   next_hop_in_ip_address = each.value.next_hop_type != "VirtualAppliance" ? null : each.value.next_hop_in_ip_address
 }
 
-resource "azurerm_subnet_route_table_association" "aks_00" {
-  route_table_id = azurerm_route_table.route_table.id
-  subnet_id      = azurerm_subnet.aks_00_subnet.id
-}
+resource "azurerm_subnet_route_table_association" "route_table" {
+  for_each = { for subnet in var.subnets : subnet.name => subnet
+    if subnet.route_table == true
```
Comment (timja) [inline 12-connectivity.tf:52]: the default should be setup to include what we expect, i.e. a route table on each subnet pointing to the hub, with an opt-out if required

🚫 Blocking

File: 12-connectivity.tf [hunk 4]
```diff
@@ -2,73 +2,17 @@
...
-  for_each = { for subnet in var.additional_subnets : subnet.name => subnet }
+resource "azurerm_subnet" "subnets" {
+  for_each = { for subnet in var.subnets : subnet.name => subnet }
 
   name                                           = each.value.name
   address_prefixes                               = [each.value.address_prefix]
   resource_group_name                            = var.resource_group_name
   virtual_network_name                           = azurerm_virtual_network.virtual_network.name
-  enforce_private_link_endpoint_network_policies = true
+  enforce_private_link_endpoint_network_policies = each.value.enforce_private_link_endpoint_network_policies == null ? false : var.iaas_subnet_enforce_private_link_endpoint_network_policies
```
Comment (timja) [inline 12-connectivity.tf:14]: why are you using the `iaas_subnet_enforce_private_link_endpoint_network_policies` var? 

don't you want this?

```suggestion
  enforce_private_link_endpoint_network_policies = each.value.enforce_private_link_endpoint_network_policies == null ? false : each.value.enforce_private_link_endpoint_network_policies
```

🚫 Blocking

File: 12-connectivity.tf [hunk 5]
```diff
@@ -2,73 +2,17 @@
...
+  service_endpoints                              = each.value.subnet_service_endpoints == null ? [] : var.subnet_service_endpoints
```
Comment (timja) [inline 12-connectivity.tf:15]:
```suggestion
  service_endpoints                              = each.value.subnet_service_endpoints == null ? [] : each.value.subnet_service_endpoints
```

**Review by timja — CHANGES_REQUESTED** [review]
This is a breaking change to the module, do you plan to migrate state state of each environment on SDS / CFT to migrate the state across?

🚫 Blocking

File: 00-init.tf (hunk 1 above)
Comment (thomast1906) [reply 00-init.tf:3]: Yeah, this was added to use optional
https://github.com/hmcts/aks-module-network/blob/subnets-foreach/01-inputs-required.tf#L22-L24 

Can just include in the .tfvars & remove this experiment

File: 02-inputs-optional.tf (hunk 2 above)
Comment (thomast1906) [reply 02-inputs-optional.tf:16]: Yeah, copy/paste fail

File: 12-connectivity.tf (hunk 3 above)
Comment (thomast1906) [reply 12-connectivity.tf:52]: Sure - will swap to `!= false`

File: 12-connectivity.tf (hunk 4 above)
Comment (thomast1906) [reply 12-connectivity.tf:14]: Was just grabbing it from the default defined in the module. Can remove and just include in my for_each.

File: 12-connectivity.tf (hunk 5 above)
Comment (thomast1906) [reply 12-connectivity.tf:15]: Was just grabbing it from the default defined in the module. Can remove and just include in my for_each.

File: 12-connectivity.tf (hunk 4 above)
Comment (thomast1906) [reply 12-connectivity.tf:14]: Added now to each.value

**Comment by thomast1906** [comment]
Closing as old

**Pull request closed without merging**